File: smalivm/src/test/java/org/cf/smalivm/opcode/SwitchPayloadOpTest.java
package org.cf.smalivm.opcode;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.mockito.Mockito.withSettings;

import gnu.trove.map.TIntObjectMap;
import gnu.trove.map.hash.TIntObjectHashMap;
import java.util.LinkedList;
import java.util.List;
import org.cf.smalivm.VirtualMachine;
import org.cf.smalivm.VirtualMachineImpl;
import org.cf.smalivm.type.ClassManager;
import org.cf.smalivm.type.ClassManagerImpl;
import org.jf.dexlib2.Opcode;
import org.jf.dexlib2.builder.BuilderInstruction;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.SwitchElement;
import org.jf.dexlib2.iface.instruction.SwitchPayload;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class SwitchPayloadOpTest {

    public static class UnitTest {

        private static final int ADDRESS = 0;

        private TIntObjectMap<MethodLocation> addressToLocation;
        private ClassManager classManager;
        private MethodLocation location;
        private SwitchPayloadOp op;
        private SwitchPayloadOpFactory opFactory;
        private BuilderInstruction instruction;
        private VirtualMachine vm;

        @Test
        public void payloadOpHasExpectedToString() {
            when(instruction.getOpcode()).thenReturn(Opcode.PACKED_SWITCH_PAYLOAD);
            List<SwitchElement> elements = new LinkedList<SwitchElement>();
            doReturn(elements).when((SwitchPayload) instruction).getSwitchElements();
            SwitchElement element1 = mock(SwitchElement.class);
            when(element1.getKey()).thenReturn(1);
            when(element1.getOffset()).thenReturn(10);
            SwitchElement element2 = mock(SwitchElement.class);
            when(element2.getKey()).thenReturn(2);
            when(element2.getOffset()).thenReturn(20);
            elements.add(element2);
            elements.add(element1);

            op = (SwitchPayloadOp) opFactory.create(location, addressToLocation, vm);

            assertEquals("packed-switch-payload [1 -> :addr_10, 2 -> :addr_20]", op.toString());
        }

        @BeforeEach
        public void setUp() {
            vm = mock(VirtualMachineImpl.class);
            classManager = mock(ClassManagerImpl.class);
            when(vm.getClassManager()).thenReturn(classManager);

            location = mock(MethodLocation.class);
            instruction = mock(BuilderInstruction.class, withSettings().extraInterfaces(SwitchPayload.class));
            when(location.getInstruction()).thenReturn(instruction);
            when(location.getCodeAddress()).thenReturn(ADDRESS);
            when(instruction.getLocation()).thenReturn(location);
            when(instruction.getCodeUnits()).thenReturn(0);

            addressToLocation = new TIntObjectHashMap<MethodLocation>();
            addressToLocation.put(ADDRESS, location);

            opFactory = new SwitchPayloadOpFactory();
        }

    }

}

File: smalivm/src/test/java/org/cf/smalivm/opcode/SwitchPayloadOpTest.kt
package org.cf.smalivm.opcode

import io.mockk.every
import io.mockk.mockk
import org.jf.dexlib2.Opcode
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.builder.instruction.BuilderSparseSwitchPayload
import org.jf.dexlib2.builder.instruction.BuilderSwitchElement
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class SwitchPayloadOpTest {
    companion object {
        private const val ADDRESS = 0
    }

    @Nested
    inner class UnitTest {
        private lateinit var location: MethodLocation
        private lateinit var instruction: BuilderSparseSwitchPayload

        @BeforeEach
        fun setUp() {
            instruction = mockk {
                every { codeUnits } returns 0
            }
            location = mockk {
                every { codeAddress } returns ADDRESS
            }
            every { location.instruction } returns instruction
            every { instruction.location } returns location
        }

        @Test
        fun payloadOpHasExpectedToString() {
            every { instruction.opcode } returns Opcode.PACKED_SWITCH_PAYLOAD

            val element1: BuilderSwitchElement = mockk {
                every { key } returns 1
                every { offset } returns 10
            }
            val element2: BuilderSwitchElement = mockk {
                every { key } returns 2
                every { offset } returns 20
            }
            val elements: MutableList<BuilderSwitchElement> = mutableListOf(element1, element2)
            every { instruction.switchElements } returns elements

            val op = SwitchPayloadOp.build(location, mockk(), mockk(), mockk())
            assertEquals(
                "packed-switch-payload [${element1.key} -> :addr_${element1.offset}, ${element2.key} -> :addr_${element2.offset}]",
                op.toString()
            )
        }
    }
}
